projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-token-subscription-service.php

```diff
@@ -1,55 +0,0 @@
-<?php
-/**
- * A paywall that exchanges JWT tokens from WordPress.com to allow
- * a current visitor to view content that has been deemed "Premium content".
- *
- * @package Automattic\Jetpack\Extensions\Premium_Content
- */
-
-namespace Automattic\Jetpack\Extensions\Premium_Content\Subscription_Service;
-
-/**
- * Class WPCOM_Token_Subscription_Service
- *
- * @package Automattic\Jetpack\Extensions\Premium_Content\Subscription_Service
- */
-class WPCOM_Token_Subscription_Service extends Token_Subscription_Service {
-
-	/**
-	 * Is available()
-	 *
-	 * @inheritDoc
-	 */
-	public static function available() {
-	 // phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
-		return defined( 'IS_WPCOM' ) && IS_WPCOM === true;
-	}
-
-	/**
-	 * Is get_site_id()
-	 *
-	 * @inheritDoc
-	 */
-	public function get_site_id() {
-		return get_current_blog_id();
-	}
-
-	/**
-	 * Is get_key()
-	 *
-	 * @inheritDoc
-	 */
-	public function get_key() {
-	 // phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
-		return defined( 'EARN_JWT_SIGNING_KEY' ) ? EARN_JWT_SIGNING_KEY : false;
-	}
```

Comment:
It looks like the tests are expecting this code to be available. In this refactor, it's hidden behind `( new Host() )->is_wpcom_simple()`. The problem is that these tests run outside of WPCOM simple, so this `is_wpcom_simple` test will always be false. I think the easy solution here is to make the tests use `WPCOM_Online_Subscription_Service` rather that `Jetpack_Token_Subscription_Service`.

Reply:
I've started exploring these approaches. I understand that some work will need to be done to refactor these tests if we truly want to make this change. I'll leave some notes here on the approaches as I work through the code on the work required so I can compare the approaches.

TL:DR; I went with `Test_Jetpack_Token_Subscription_Service`, and explained how I came to that conclusion.

**Problem**
Tests which assert that a user can access content under specific conditions are now reporting `false` when `true` was expected. This is due to an early `return false` statement because a key was not returned from `get_key`. We just need to return a string for these tests to pass.

**Mocking `Host`**
Actually mocking `Host` here might be difficult, since the test doesn't define which files get included and `Host` is already present. I checked to see how `Host` is tested, and it seems it uses approach 2 via `Constants::set_constant`.

**Setting `IS_WPCOM`** 
I assume with approach 2 that you mean setting the constant via `define( `IS_WPCOM`, true )`; if so I agree that this is not a good idea. However, `Host` can be manipulated to do what we want with functions like `Constants::set_constant()`, since `Host` uses the testable `Constants` class. I can use in tests, but doing so presents a new problem (one we'd still face even if we could easily mock `Host`): the tests weren't expecting to need to call `is_jetpack_site()` (which doesn't exist) but they now are.

We only want to set `IS_WPCOM` at a certain depth - just enough to get that key and then we want to clear the constant. After all, this is all being done just to get `get_key`, which in `WPCOM_Token_Subscription_Service` is simply returned when `get_key` is called. Now that we have merged the two, we need to short-cut just enough logic (`( new Host() )->is_wpcom_simple() && is_jetpack_site()`) in only `Jetpack_Token_Subscription_Service` and nowhere else. This is a difficult task, because outside of this one specific case where we `get_key`, all other calls to functions called by functions under test need to be run as if `IS_WPCOM` is not defined.

**Refactor the code to early return in certain conditions based on `TESTING_IN_JETPACK`**
Remember that the end goal here is to return `defined( 'EARN_JWT_SIGNING_KEY' ) ? EARN_JWT_SIGNING_KEY : false`. This code (inside `get_key`) is behind `if ( ( new Host() )->is_wpcom_simple() ) {`:

```
	public function get_key() {
		if ( ( new Host() )->is_wpcom_simple() ) {
			// phpcs:ignore ImportDetection.Imports.RequireImports.Symbol
			return defined( 'EARN_JWT_SIGNING_KEY' ) ? EARN_JWT_SIGNING_KEY : false;
		}
		$token = ( new Tokens() )->get_access_token();
		if ( ! isset( $token->secret ) ) {
			return false;
		}
		return $token->secret;
	}
```

I can add a `( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) || ` before the check for `is_wpcom_simple`. This would fix the issue for this test, but then I figured it should present an issue when testing `Jetpack_Token_Subscription_Service` directly. However, when I searched for this specific class under test inside tests, I found `Test_Jetpack_Token_Subscription_Service` which `extends Jetpack_Token_Subscription_Service`.

I ran the phpunit tests for this, and they passed. However, knowing now about `Test_Jetpack_Token_Subscription_Service`, and another approach we could take is..

**Mock/Double Jetpack_Token_Subscription_Service**
This class extends the class we want to test, overwriting the `get_key` function to just return a random string, which is all we need for testing. This isn't a very good test if `Jetpack_Token_Subscription_Service` because nothing in that function is being tested. At least with `WPCOM_Token_Subscription_Service` was used, the `EARN_JWT_SIGNING_KEY` would be returned. This means everything  from `$token = ( new Tokens() )->get_access_token();` and below isn't being tested, but it also means that it never was covered in tests to begin with.

This has the downside of needing to change the tests to test `Test_Jetpack_Token_Subscription_Service`, but I those are changes to tests, and not changes to code, so I think that's better.

Reply:
> Mock/Double Jetpack_Token_Subscription_Service

> but it also means that it never was covered in tests to begin with.
I don't think it was ever the point of testing this particular code. So all good.

I think it is a good approach.